app/javascript/app/document-capture/hooks/use-async.js

```diff
@@ -0,0 +1,69 @@
+import { useState, useCallback } from 'react';
+import useIfStillMounted from './use-if-still-mounted';
+
+/**
+ * @typedef SuspenseResource
+ *
+ * @prop {()=>any} read Resource reader, called from a descendent of a Suspense
+ *                      element. The reader will either throw an error, throw
+ *                      the pending promise, or return the value of the resolved
+ *                      promise upon completion.
+ */
+
+/**
+ * Given a function which returns a promise, returns a Suspense resource object.
+ * The resource object can be read from a descendent of a Suspense element,
+ * allowing for fallback states for loading or error handling. Any additional
+ * arguments are passed through to the promise creator, and are treated as
+ * dependencies to trigger a new promise to be issued.
+ *
+ * @see https://reactjs.org/docs/concurrent-mode-suspense.html
+ *
+ * @param {(...args:any)=>Promise} createPromise Promise creator.
+ * @param {...any}                 args          Additional arguments to pass to
+ *                                               promise creator.
+ *
+ * @return {SuspenseResource} Suspense resource object.
+ */
+function useAsync(createPromise, ...args) {
+  const ifStillMounted = useIfStillMounted();
+  const [data, setData] = useState();
+  const [hasData, setHasData] = useState(false);
+  const [error, setError] = useState();
+  const [hasError, setHasError] = useState(false);
+  const read = useCallback(
+    () =>
+      createPromise(...args)
+        .then(
+          ifStillMounted((nextData) => {
+            setData(nextData);
+            setHasData(true);
+          }),
+        )
+        .catch(
+          ifStillMounted((nextError) => {
+            setError(nextError);
+            setHasError(true);
+          }),
+        ),
+    args,
+  );
+
+  return {
+    read() {
+      const promise = read();
+
+      if (hasData) {
+        return data;
+      }
+
+      if (hasError) {
+        throw error;
+      }
+
+      throw promise;
```

Comment:
Dumb question....how come this isn't `return`? docs up top say:

> or return the value of the resolved promise upon completion.

Reply:
The quoted text refers to the path above where `hasData` condition will `return data;`. This line in particular is documented on the prior line:

>throw the pending promise

I might be getting a little ahead of myself by venturing into the bleeding-edge territory of [React Suspense](https://reactjs.org/docs/concurrent-mode-suspense.html), but the primary thing that Suspense introduces is the idea that you can throw a `Promise`, and React will treat that as a pending state, allowing the developer to render a `fallback` specified by a parent `Suspense` element.

So there are three expected possible states:

- Throw a promise: Pending
- Throw an error: Error
- Return a value: Complete

Reply:
Ah I didn't RTFM well enough.... thanks for explaining!